server/bin/pbench-tarball-stats

```diff
@@ -0,0 +1,178 @@
+    """
+    with os.scandir(archive) as archive_scan:
+        for c_entry in archive_scan:
+            if c_entry.name.startswith(".") or not c_entry.is_dir(follow_symlinks=False):
+                continue
+            # We have a controller directory.
+            with os.scandir(c_entry.path) as controller_scan:
+                for entry in controller_scan:
+                    if entry.name.startswith(".") or not entry.is_file():
+                        continue
+                    match = tb_pat.fullmatch(entry.name)
+                    if match:
+                        # Turn the pattern components of the match into a datetime
+                        # object.
+                        tb_dt = datetime(
+                            int(match.group(1)),
+                            int(match.group(2)),
+                            int(match.group(3)),
+                            int(match.group(4)),
+                            int(match.group(5)),
+                            int(match.group(6)),
+                        )
+                    else:
+                        if entry.name.endswith(".md5"):
+                            continue
+                        tb_dt = None
+                    yield c_entry.name, entry.name, tb_dt, entry.stat()
+
+
+def report_by_bucket(buckets):
+    for bucket_name, bucket_list in sorted(buckets.items()):
+        count = len(bucket_list)
+        ec2_cnt = 0
+        for item in bucket_list:
+            if item[0].startswith("EC2::"):
+                ec2_cnt += 1
+        pct_ec2 = (ec2_cnt / count) * 100.0
+        print(f"{bucket_name}: {count:7n} (%EC2 {pct_ec2:0.02f})")
+
+
+def main(options):
+    locale.setlocale(locale.LC_ALL, "en_US.UTF-8")
+    if not options.cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set"
+            " _PBENCH_SERVER_CONFIG env variable",
+            file=sys.stderr,
+        )
+        return 1
+
+    try:
+        config = PbenchConfig(options.cfg_name)
+    except BadConfig as e:
+        print(f"{_NAME_}: {e}", file=sys.stderr)
+        return 2
+
+    logger = get_pbench_logger(_NAME_, config)
+
+    archive = config.ARCHIVE
+    archive_p = os.path.realpath(archive)
+
+    if not archive_p:
+        logger.error("The configured ARCHIVE directory, {}, does not exist", archive)
+        return 3
+
+    if not os.path.isdir(archive_p):
+        logger.error(
+            "The configured ARCHIVE directory, {}, is not a valid directory", archive
+        )
+        return 4
+
+    invalid_cnt = 0
+    invalid_size = 0
+    invalid = collections.defaultdict(list)
+
+    good_cnt = 0
+    good_size = 0
+    good = collections.defaultdict(list)
+
+    ec2_cnt = 0
+    rh_cnt = 0
+
+    by_year = collections.defaultdict(list)
+    by_month = collections.defaultdict(list)
+
+    start = pbench._time()
+
+    gen = gen_tar_balls(archive_p)
+
+    for tb_rec in gen:
```

Comment:
It would be better if you used a `namedtuple` or if you destructured the tuple into separately-named local variables, here.